File: client/publisher.go
package client

import (
	"github.com/gtfierro/cs262-project/common"
	"sync"
	"sync/atomic"
)

type Publisher struct {
	BrokerConnection
	ID common.UUID

	// metadata for this client if it acts as a publisher
	metadataLock sync.RWMutex
	// if true, then there is metadata this client needs to send
	// to the broker in its next publish message
	dirtyMetadata bool
	metadata      map[string]interface{}

	sentMessagesAttempt uint32
	sentMessagesTotal   uint32
}

// publishLoop should be a loop which will start whenever a connection to a broker
// is established; it should exit whenever an error is encountered and it will be
// restarted once a reconnection is reestablished
func NewPublisher(id common.UUID, publishLoop func(*Publisher), cfg *Config) (p *Publisher, err error) {
	p = &Publisher{
		ID:                  id,
		metadata:            make(map[string]interface{}),
		sentMessagesAttempt: 0,
		sentMessagesTotal:   0,
	}
	connectCallback := func() {
		go publishLoop(p)
	}
	err = (&p.BrokerConnection).initialize(connectCallback, func(common.Sendable) {}, cfg)
	if err != nil {
		p = nil
		return
	}

	return
}

func (p *Publisher) Publish(value interface{}) error {
	atomic.AddUint32(&p.sentMessagesAttempt, 1)
	m := &common.PublishMessage{
		UUID:  p.ID,
		Value: value,
	}
	p.metadataLock.RLock()
	if p.dirtyMetadata {
		m.Metadata = p.metadata
	}
	p.dirtyMetadata = false
	p.metadataLock.RUnlock()
	// if we fail to send, then reconnect
	if err := p.sendBroker(m); err != nil {
		return err
	}
	atomic.AddUint32(&p.sentMessagesTotal, 1)
	return nil
}

func (p *Publisher) GetStats() PublisherStats {
	stats := PublisherStats{
		MessagesAttempted:  atomic.LoadUint32(&p.sentMessagesAttempt),
		MessagesSuccessful: atomic.LoadUint32(&p.sentMessagesTotal),
	}
	return stats
}

func (p *Publisher) AddMetadata(newm map[string]interface{}) {
	// TODO: do we need to do anything beyond a blind overwrite?
	p.metadataLock.Lock()
	for k, v := range newm {
		// deletes are handled by making v nil
		p.metadata[k] = v
	}
	p.dirtyMetadata = true
	p.metadataLock.Unlock()
}

type PublisherStats struct {
	MessagesAttempted  uint32
	MessagesSuccessful uint32
}

File: python_client/publisher.py
from broker_connection import BrokerConnection
from messages import PublishMessage
import threading


class Publisher:

    # publish_loop should be a loop which will start whenever a connection
    # to the broker is established; it should exit whenever an error is encountered
    # and will be restarted once a connection is reestablished
    # receives this publisher as an argument
    def __init__(self, uuid, publish_loop, config):
        self.uuid = uuid
        self.metadata = {}
        self.dirty_md = False
        self.md_lock = threading.Lock()
        self.publish_loop = publish_loop

        self.broker_conn = BrokerConnection(config, self.connect_callback, None, uuid)

    def start(self):
        self.broker_conn.start()

    def connect_callback(self):
        publish_loop_thread = threading.Thread(target=self.publish_loop_wrap)
        publish_loop_thread.start()

    def publish_loop_wrap(self):
        self.publish_loop(self)

    def publish(self, value):
        metadata = {}
        self.md_lock.acquire()
        if self.dirty_md:
            metadata = self.metadata
            self.dirty_md = False
        self.md_lock.release()

        msg = PublishMessage(self.uuid, value, metadata)
        self.broker_conn.send(msg)

    def add_metadata(self, new_metadata):
        self.md_lock.acquire()
        for k, v in new_metadata.iteritems():
            self.metadata[k] = v
        self.dirty_md = True
        self.md_lock.release()
